src/component.cpp

```diff
@@ -151,24 +151,21 @@ void Component::removeAllVariables()
     mPimpl->mVariables.clear();
 }
 
-VariablePtr Component::getVariable(size_t index)
+VariablePtr Component::getVariable(size_t index) const
 {
-    return mPimpl->mVariables.at(index);
-}
-
-const VariablePtr& Component::getVariable(size_t index) const
-{
-    return mPimpl->mVariables.at(index);
-}
-
-VariablePtr Component::getVariable(const std::string &name)
-{
-    return mPimpl->mVariables.at(mPimpl->findVariable(name) - mPimpl->mVariables.begin());
+    if (index < mPimpl->mVariables.size()) {
+        return mPimpl->mVariables.at(index);
+    }
+    return nullptr;
 }
 
-const VariablePtr& Component::getVariable(const std::string &name) const
+VariablePtr Component::getVariable(const std::string &name) const
 {
-    return mPimpl->mVariables.at(mPimpl->findVariable(name) - mPimpl->mVariables.begin());
+    auto result = mPimpl->findVariable(name);
+    if (result == mPimpl->mVariables.end()) {
+        return nullptr;
+    }
+    return *result;
```

Comment:
Minor: for consistency with some other bits of the code (at least), it might be better to have (same for `UnitsPtr ComponentEntity::getUnits(const std::string &name) const`):

```
auto result = mPimpl->findVariable(name);
if (result != mPimpl->mVariables.end()) {
    return *result;
}
return nullptr;
```

i.e. finish with `return nullptr;`.